src/sentry/dynamic_sampling/latest_release_booster.py

```diff
@@ -1,234 +1,382 @@
-            for boosted_release in self.boosted_releases
-            if boosted_release.id in models
-        ]
+        models = self._get_releases_models()
 
-    def _get_last_release_ids(self, limit: int) -> List[int]:
-        return [boosted_release.id for boosted_release in self.boosted_releases[-limit:]]
+        current_timestamp = datetime.utcnow().replace(tzinfo=UTC).timestamp()
 
-    def _get_releases_models(self, limit: int) -> Dict[int, Release]:
+        extended_boosted_releases = []
+        expired_boosted_releases = []
+        for boosted_release in self.boosted_releases:
+            extended_boosted_release = boosted_release.extend(
+                release=models[boosted_release.id], project_id=project_id
+            )
+
+            if extended_boosted_release.is_active(current_timestamp):
+                extended_boosted_releases.append(extended_boosted_release)
+            else:
+                expired_boosted_releases.append(boosted_release.cache_key)
+
+        return extended_boosted_releases, expired_boosted_releases
+
+    def _get_last_release_ids(self) -> List[int]:
+        return [boosted_release.id for boosted_release in self.boosted_releases]
+
+    def _get_releases_models(self) -> Dict[int, Release]:
         return {
             release.id: release
-            for release in Release.objects.filter(id__in=self._get_last_release_ids(limit=limit))
+            for release in Release.objects.filter(id__in=self._get_last_release_ids())
         }
 
 
-def _get_environment_cache_key_suffix(environment: Optional[str]) -> str:
-    return f"{ENVIRONMENT_SEPARATOR}{environment}" if environment else ""
+class ProjectBoostedReleases:
+    """
+    Class responsible of hiding the complexity of handling boosted releases in the Redis hash. In addition, it provides
+    all the logic to handle an upper bound in the number of boosted releases that can be simultaneously be added to
+    a specific project.
+    """
 
+    # Limit of boosted releases per project.
+    BOOSTED_RELEASES_LIMIT = 10
+    BOOSTED_RELEASES_HASH_EXPIRATION = 60 * 60 * 1000
+
+    def __init__(self, project_id: int):
+        self.redis_client = get_redis_client_for_ds()
+        self.project_id = project_id
+        self.project_platform = _get_project_platform(self.project_id)
+
+    def add_boosted_release(self, release_id: int, environment: Optional[str]) -> None:
+        """
+        Adds a release to the boosted releases hash with the boosting timestamp set to the current time, signaling that
+        the boosts starts now.
+        """
+        self._remove_lrb_if_limit_is_reached()
+
+        cache_key = self._generate_cache_key_for_boosted_releases_hash()
+        self.redis_client.hset(
+            cache_key,
+            self._generate_cache_key_for_boosted_release(release_id, environment),
+            datetime.utcnow().replace(tzinfo=UTC).timestamp(),
```

Comment:
Not blocking, but I would store the expiry time in redis, instead of the current time. That way, you only need `time_to_adoption` when you write, and all reads can simply check `current_timestamp < expiry_timestamp`.

Reply:
This was under consideration but for flexibility we decided to actually store the current, so that we can adapt and change it later on without breaking production.

Reply:
Keep in mind that this implementation will be kept for a few weeks because I will do another one from scratch with the dynamic approach. This was mainly a small step to verify some ideas.